Added logic to download opensearch xml and load engine

danishjafri88 opened the pull request:
This pr includes code to allow user add a custom search engine. This is possible in two ways:
1. Add OpenSearch engine
    - From BVC by selecting a search field, an option to add search engine will appear above the keyboard.
    - From search settings while adding a search engine, if the site supports OpenSearch and option to add it will be provided in the header.

2. Manually
    User can add an engine manually in settings by entering a URL with formatted search query param and a title.

![Simulator Screen Shot - iPhone X - 2019-03-29 at 22 36 52](https://user-images.githubusercontent.com/5398557/55250015-5d354680-5273-11e9-8f78-1b48bb338206.png)
![Simulator Screen Shot - iPhone X - 2019-03-29 at 22 36 58](https://user-images.githubusercontent.com/5398557/55250027-62929100-5273-11e9-9f1f-caa2a1399c31.png)
![Simulator Screen Shot - iPhone X - 2019-03-29 at 22 37 00](https://user-images.githubusercontent.com/5398557/55250034-64f4eb00-5273-11e9-8877-3c3013f0b5fd.png)
![Simulator Screen Shot - iPhone X - 2019-03-29 at 22 37 24](https://user-images.githubusercontent.com/5398557/55250041-66beae80-5273-11e9-87f2-c0600a6cc09d.png)
![Simulator Screen Shot - iPhone X - 2019-03-29 at 22 37 26](https://user-images.githubusercontent.com/5398557/55250046-67efdb80-5273-11e9-90a6-39acd1aee254.png)
![Simulator Screen Shot - iPhone X - 2019-03-29 at 22 37 48](https://user-images.githubusercontent.com/5398557/55250049-69b99f00-5273-11e9-8a93-d0df8556fbe7.png)
![Simulator Screen Shot - iPhone X - 2019-03-29 at 22 37 51](https://user-images.githubusercontent.com/5398557/55250052-6b836280-5273-11e9-94a9-570e23533396.png)

<!-- 
*Thank you for submitting a pull request, your contributions are greatly appreciated!*

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. 
List any dependencies that are required for this change.
-->

## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [ ] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [ ] *Unit Tests* are updated to cover new or changed functionality
- [ ] User-facing strings use `NSLocalizableString()`
- [ ] New files have MPL-2.0 license header.


## Test Plan:

<!-- Any useful notes for reviewer explaining how best to test and verify. -->

### Screenshots:

<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [ ] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [ ] Issues are assigned to at least one epic.
- [ ] Issues include necessary QA labels:
  - [ ] `QA/(Yes|No)`
  - [ ] `release-notes/(include|exclude)`
  - [ ] `bug` / `enhancement`
- [ ] Necessary security reviews have taken place.
- [ ] Adequate test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable)

danishjafri88 committed:
Added logic to download opensearch xml and load engine

JonKaleidoscope commented on Shared/Extensions/StringExtensions.swift:
```diff
@@ -109,4 +109,15 @@ public extension String {
+            if let data = self.data(using: .utf8) {
+                return try JSONSerialization.jsonObject(with: data, options: .mutableContainers)
+            }
+        } catch {}
```
Since you are not using the catch you could refactor this function to use a `try?` If an error is thrown it will just return `nil`

```swift
guard let data = self.data(using: .utf8) else { return nil }
return try? JSONSerialization.jsonObject(with: data, options: .mutableContainers)
```

danishjafri88 replied:
Hi, thanks reviewing this pr. Unfortunately this pr is still WIP and in draft, there are lot more changes that will be pushed out soon. 
In this case I have added a logger, hence we need to keep the implementation.

JonKaleidoscope replied:
Gotcha

JonKaleidoscope commented on Shared/Extensions/StringExtensions.swift:
```diff
@@ -109,4 +109,15 @@ public extension String {
+    public func jsonObject() -> Any? {
+        do {
+            if let data = self.data(using: .utf8) {
+                return try JSONSerialization.jsonObject(with: data, options: .mutableContainers)
```
I think you will be fine without the `.mutableContainers` since you are calling this function in swift. The containers are mutable by default.

danishjafri88 replied:
Updated thanks!

JonKaleidoscope commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2592,52 +2593,51 @@ extension BrowserViewController: HistoryStateHelperDelegate {
+                //check if the search engine has already been added.
+                self.openSearchLinkDict = dict
+                let shortName = dict["title"] ?? ""
+                let domain = 
```
Is there no value for domain?

danishjafri88 replied:
Hehe well it was committed in a hurry, this is history now.

JonKaleidoscope commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2592,52 +2593,51 @@ extension BrowserViewController: HistoryStateHelperDelegate {
-                make.height.equalTo(inputView.snp.height)
+        self.openSearchLinkDict = nil
+        webView.evaluateJavaScript(OpenSearchEngine.fetchOpenSearchLinkScript) { (result, _) in
+            if let dict = (result as? String)?.jsonObject() as? [String: String] {
```
I think a `guard let` might be better served here. 

Also you might get away with using a JSON Decoder here since strings automatically conform to `Cobale`.
```swift
let decoder = JSONDecoder()
guard let dict = try? decoder.decode([String: String].self, from: result) else { return }
```

danishjafri88 replied:
The problem is `result` is of type `Any` while the expected type is String(the script returns a string). The Decoder requires a `Data` type. Hence the helper function. Since the function is in String Extension I cannot hardcode it to `[String: String]` type.

JonKaleidoscope replied:
Fair enough.

danishjafri88 committed:
Added check for relative paths in open search link href.

danishjafri88 committed:
Added Open Search option to Custom Search Engine Add Screen

danishjafri88 committed:
Logic to add Open search engine refactor end 2 end

danishjafri88 committed:
Added logic to add search engine manually

danishjafri88 commented on Client/Utils/FaviconFetcher.swift:
```diff
@@ -147,26 +146,43 @@ open class FaviconFetcher: NSObject, XMLParserDelegate {
-                // If we haven't got any options icons, then use the default at the root of the domain.
-                if let url = NSURL(string: "/favicon.ico", relativeTo: url as URL), icons.isEmpty, let absoluteString = url.absoluteString {
+                
+                // SVG is not supported right not
```
Are SVG images supported or am I missing something? Should we add support for svg icons. Maybe in a new PR?

danishjafri88 committed:
Changed the UI to show:
1. What Search Engine is added in Auto Add
2. Show Auto Add button in header
3. Show loader in nav bar.

danishjafri88 committed:
Merge branch 'development' into feature.openSearch.936

danishjafri88 committed:
Fixed build conflict with development

danishjafri88 committed:
Merge branch 'development' into feature.openSearch.936

jhreis commented:
When this is merged, please close https://github.com/brave/brave-ios/issues/1030

jhreis committed:
Merge branch 'development' into feature.openSearch.936

jhreis committed:
Merge branch 'development' into feature.openSearch.936

jhreis committed:
Fixed merging issues.
